### ISSUE TITLE
Bad performance for EK-100 Action Recognition finetuning model

### ISSUE THREAD

**Issue by HaozheQi** [zhaoyue-zephyrus/AVION#9]
Dear Yue,

Thanks for this amazing repo, which is well-documented and easy to try on! However, when I followed your guidance to fine-tune the action recognition model on EK-100. I got very bad results (`Acc@1 0.021 Acc@5 0.089` for test and `train_Acc@1": 2.62106393129771, "train_Acc@5": 9.759661259541986` even for training). Do you have any idea about that?

To include more details for your reference, I used the resized, chunked version of EK-100 that you shared [here](https://utexas.box.com/v/ek100-320p-15sec-30fps). I used the pre-trained model you provided [here](https://utexas.app.box.com/s/e681nrxivc9makufvrumrfuaopk57h4n). And the following is the command I used for fine-tuning.

```bash
PYTHONPATH=.:third_party/decord/python/ \
CUDA_VISIBLE_DEVICES="0,1" \
torchrun \
    --nproc_per_node=2 scripts/main_lavila_finetune_cls.py \
    --root datasets/EK100/EK100_320p_15sec_30fps_libx264/ \
    --video-chunk-length 15 --use-flash-attn \
    --grad-checkpointing \
    --use-fast-conv1 \
    --batch-size 256 \
    --fused-decode-crop \
    --use-multi-epochs-loader \
    --pretrain-model experiments/pretrain_lavila_vitb/avion_pretrain_lavila_vitb_best.pt \
    --output-dir experiments/EK100_test 2>&1 | tee experiments/EK100_test/train_log.txt
```

And I also attached the generated [train_log.txt](https://github.com/zhaoyue-zephyrus/AVION/files/15235579/train_log.txt) for your reference.

Thank you in advance for your help!

**Comment by HaozheQi** [comment]
### Update
I would like to share more of my trials for your reference.

- Firstly, since the training loss cannot converge to a reasonable performance, I tried to train based on only one sample. It showed the model can achieve 100% accuracy, which means the back propagation actually works.
- Secondly, I also tried to train the model with different backbones and different machines. It seems to show that they all converge to the same local minima, as is shown in the following table.

![image](https://github.com/zhaoyue-zephyrus/AVION/assets/32994368/3d060d35-d046-4ac6-8527-ff75d2ceb23b)

**Comment by zhaoyue-zephyrus** [comment]
Hey @HaozheQi ,

The config seems okay to me. Could you try one with the same per-GPU batch size? We use a default batch size of 32. Not sure if something wrong was triggered because of that. You can also check out our training log [here](https://utexas.box.com/s/ecxop240sin3ur1hevb4w9tcvwniiabv).

**Comment by FranklinLeong** [comment]
Dear Yue,

Firstly, thank you for the repo as well, it was super interesting and we're also trying to use the model for our purpose. As a starter, I am just trying to fine tune the model on the EK100 dataset before testing on my own dataset. However, I am also facing similar issue to what Qi had suggested. this was the config that I used to train on the full dataset with batch-size 32. 

 `PYTHONPATH=.:third_party/decord/python/ torchrun \
    --nproc_per_node=1 scripts/main_lavila_finetune_cls.py \
    --root datasets/EK100/EK100_320p_15sec_30fps_libx264/ \
    --video-chunk-length 15 --use-flash-attn \
    --grad-checkpointing \
    --use-fast-conv1 \
    --batch-size 32 \
    --fused-decode-crop \
    --use-multi-epochs-loader \
    --pretrain-model experiments/pretrain_lavila_vitb/avion_pretrain_lavila_vitb_best.pt \
    --output-dir test_fine_tuning 2>&1 | tee test_fine_tuning/log.txt`

This was what I got after 4 epochs (unfortunately, something went wrong during the training and crashed my computer). 

![image](https://github.com/zhaoyue-zephyrus/AVION/assets/51982615/914662f4-46f8-4939-a503-725d64f2e185)

Compared to your training log, it seems already much lower even after the same number of epochs.
![image](https://github.com/zhaoyue-zephyrus/AVION/assets/51982615/e5ba4417-587f-4255-9bf2-b02e01c2df79)

Next, I tried training with only 100 samples to simplify the problem and tried getting it to overfit, but it did not manage to overfit even on the training set as shown below.

![image](https://github.com/zhaoyue-zephyrus/AVION/assets/51982615/6e3bae19-1c2d-4aee-9528-7fc55ea08918)

I would be grateful if you could provide any sort of advice to the problem.

**Comment by zhaoyue-zephyrus** [comment]
Hi @FranklinLeong ,

 I am not sure if @HaozheQi solved the issue but during the last email communication with him, I mentioned trying to turn off the [loss scale](https://github.com/zhaoyue-zephyrus/AVION/blob/main/scripts/main_lavila_finetune_cls.py#L231C46-L231C57) to see if it works. For bf16 loss scaling is [not required](https://discuss.pytorch.org/t/why-bf16-do-not-need-loss-scaling/176596).

**Comment by HaozheQi** [comment]
I unfortunately haven't solved the problem, disabling the mixup and upgrading the flash-attention to flash-attention 2 can improve the performance a bit but I cannot reach the ideal performance.

**Comment by zhaoyue-zephyrus** [comment]
@HaozheQi 

Can you check the decord version by running the following command (in the same working environment): 

```
python -c "import decord; print(decord.__path__)"
```

It would be good to see if the correct decord is being used during training. The main branch would probably fail.

**Comment by HaozheQi** [comment]
Hi, @zhaoyue-zephyrus 
- Directly run `python -c "import decord; print(decord.__path__)"` would point to another path `['/home/haozhe/.local/lib/python3.10/site-packages/decord-0.6.0-py3.10-linux-x86_64.egg/decord']`
- However, when running Avion, I will run as `PYTHONPATH=.:third_party/decord/python/ python -c "import decord; print(decord.__path__)"`, this will point to the correct path. `['/home/haozhe/kitchen/AVION/third_party/decord/python/decord']`

Therefore, I think my discord installation should have no problem, I also try to plot the image in the end of the `video_loader` and the generated video images look reasonable.
![image](https://github.com/zhaoyue-zephyrus/AVION/assets/32994368/7e428741-120e-4345-96de-ac72ac3a0eb8)
![image](https://github.com/zhaoyue-zephyrus/AVION/assets/32994368/c4780cdd-6d14-4234-833e-3acc4c6b93d2)

**Comment by rishik18** [comment]
> Hey @HaozheQi ,
> 
> The config seems okay to me. Could you try one with the same per-GPU batch size? We use a default batch size of 32. Not sure if something wrong was triggered because of that. You can also check out our training log [here](https://utexas.box.com/s/ecxop240sin3ur1hevb4w9tcvwniiabv).

Hello @zhaoyue-zephyrus,
Thank you for sharing the code and the training logs.

Hi @HaozheQi,
I faced similar issues and I believe the reason is that the default value of weight decay is 0.05 whereas the the one in the training log is 4e-5. The relatively high default weight decay seems to have affected the training. Setting the weight decay to 4e-5 and changing other parameters to the ones from the log resolved the issue for me.

**Comment by zhaoyue-zephyrus** [comment]
Hi @rishik18 

Thanks for spotting this issue! I'll update the example command accordingly. 

Best,
Yue

**Comment by zhaoyue-zephyrus** [comment]
Hi @HaozheQi @FranklinLeong ,

The default optimizer uses SGD (instead of AdamW during pre-training) so a smaller weight decay value is required. Please give it another try. I'm closing this issue and feel free to re-open if the problem still exists.
